File: app/src/main/java/com/iceteaviet/fastfoodfinder/activity/main/SettingActivity.java
package com.iceteaviet.fastfoodfinder.activity.main;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SwitchCompat;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.iceteaviet.fastfoodfinder.BuildConfig;
import com.iceteaviet.fastfoodfinder.R;
import com.iceteaviet.fastfoodfinder.model.store.Store;
import com.iceteaviet.fastfoodfinder.model.store.StoreDataSource;
import com.iceteaviet.fastfoodfinder.network.FirebaseClient;
import com.iceteaviet.fastfoodfinder.ui.fragment.profile.StoreFilterDialogFragment;

import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SettingActivity extends AppCompatActivity {

    public static final String KEY_LANGUAGE = "lang";
    @BindView(R.id.tv_setting_share_app)
    TextView txtShareApp;
    @BindView(R.id.tv_setting_change_metric)
    TextView txtChangeMetric;
    @BindView(R.id.tv_setting_edit_profile)
    TextView txtEditProfile;
    @BindView(R.id.tv_setting_change_password)
    TextView txtChangePassword;
    @BindView(R.id.tv_setting_change_email)
    TextView txtChangeEmail;
    @BindView(R.id.tv_setting_notification)
    TextView txtSetNotification;
    @BindView(R.id.tv_setting_notification_email)
    TextView txtSetEmailNotification;
    @BindView(R.id.ll_setting_update_db)
    LinearLayout layoutUpdateDb;
    @BindView(R.id.progress_bar_update_db)
    ProgressBar progressBarUpdateDb;
    @BindView(R.id.iv_update_db)
    ImageView imageUpdateDb;
    @BindView(R.id.tv_setting_about_app)
    TextView txtAboutApp;
    @BindView(R.id.tv_setting_rate_app)
    TextView txtRateApp;
    @BindView(R.id.tv_setting_feedback)
    TextView txtFeedBack;
    @BindView(R.id.tv_setting_privacy_policy)
    TextView txtPrivacyPolicy;
    @BindView(R.id.tv_setting_term_of_use)
    TextView txtTermOfUse;
    @BindView(R.id.tv_setting_sign_out)
    TextView txtSignOut;
    @BindView(R.id.sw_languages)
    SwitchCompat swChangeLanguage;
    @BindView(R.id.tv_setting_english)
    TextView tvSettingLanguage;
    private boolean isVietnamese = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_setting);

        ButterKnife.bind(this);

        final SharedPreferences pref = this.getSharedPreferences(
                BuildConfig.APPLICATION_ID, Context.MODE_PRIVATE);

        isVietnamese = pref.getBoolean(KEY_LANGUAGE, false);

        if (!isVietnamese) {
            swChangeLanguage.setChecked(true);
        }
        // Initialize Firebase Auth
        if (FirebaseClient.getInstance().getAuth() == null
                || FirebaseClient.getInstance().getAuth().getCurrentUser() == null
                || (FirebaseClient.getInstance().getAuth().getCurrentUser() != null && FirebaseClient.getInstance().getAuth().getCurrentUser().isAnonymous())) {
            txtSignOut.setEnabled(false);
        }

        txtSignOut.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (FirebaseClient.getInstance().getAuth() != null) {
                    FirebaseClient.getInstance().signOut();
                    Intent intent = new Intent(SettingActivity.this, LoginActivity.class);
                    startActivity(intent);
                    finish();
                }
            }
        });

        swChangeLanguage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (isVietnamese) {
                    swChangeLanguage.setChecked(true);
                    isVietnamese = false;
                    loadLanguage("vi");

                } else {
                    loadLanguage("en");
                    swChangeLanguage.setChecked(false);
                    isVietnamese = true;
                }

                SharedPreferences.Editor editor = pref.edit();
                editor.putBoolean(KEY_LANGUAGE, isVietnamese);
                editor.apply();
            }
        });

        tvSettingLanguage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                swChangeLanguage.setChecked(true);
                if (isVietnamese) {
                    swChangeLanguage.setChecked(true);
                    isVietnamese = false;
                    loadLanguage("vi");
                } else {
                    loadLanguage("en");
                    swChangeLanguage.setChecked(false);
                    isVietnamese = true;
                }
            }
        });

        txtSetNotification.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                FragmentManager fm = getSupportFragmentManager();
                StoreFilterDialogFragment dlg = StoreFilterDialogFragment.newInstance();
                dlg.show(fm, "dialog-filter");
            }
        });

        layoutUpdateDb.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                FirebaseClient.getInstance().readDataFromFirebase(SettingActivity.this, new FirebaseClient.OnGetDataListener() {
                    @Override
                    public void onStart() {
                        imageUpdateDb.setVisibility(View.GONE);
                        progressBarUpdateDb.setVisibility(View.VISIBLE);
                    }

                    @Override
                    public void onSuccess(List<Store> data) {
                        StoreDataSource.setData(data);
                        Toast.makeText(SettingActivity.this, R.string.update_database_successfull, Toast.LENGTH_SHORT).show();
                        imageUpdateDb.setVisibility(View.VISIBLE);
                        progressBarUpdateDb.setVisibility(View.GONE);
                    }

                    @Override
                    public void onFailed(String errorMessage) {
                        Toast.makeText(SettingActivity.this, R.string.update_database_failed + errorMessage, Toast.LENGTH_SHORT).show();
                        imageUpdateDb.setVisibility(View.VISIBLE);
                        progressBarUpdateDb.setVisibility(View.GONE);
                    }
                });
            }
        });
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        // refresh your views here
        super.onConfigurationChanged(newConfig);
    }

    public void loadLanguage(String lang) {
        String languageToLoad = lang;
        Locale locale = new Locale(languageToLoad);
        Locale.setDefault(locale);
        Configuration configuration = new Configuration();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            setSystemLocale(configuration, locale);
        } else {
            setSystemLocaleLegacy(configuration, locale);
        }

        getBaseContext().getResources().updateConfiguration(configuration,
                getBaseContext().getResources().getDisplayMetrics());
        changeLanguage();
    }

    @SuppressWarnings("deprecation")
    public void setSystemLocaleLegacy(Configuration config, Locale locale) {
        config.locale = locale;
    }

    @TargetApi(Build.VERSION_CODES.N)
    public void setSystemLocale(Configuration config, Locale locale) {
        config.setLocale(locale);
    }

    public void changeLanguage() {
        txtShareApp.setText(R.string.share_app_with_friends);
        txtChangeMetric.setText(R.string.use_metric_units);
        txtEditProfile.setText(R.string.edit_your_profile);
        txtChangePassword.setText(R.string.change_your_password);
        txtChangeEmail.setText(R.string.change_your_email);
        txtSetNotification.setText(R.string.set_notifications);
        tvSettingLanguage.setText(R.string.english);
        txtAboutApp.setText(R.string.about_fastfood_finder);
        txtRateApp.setText(R.string.rate_app);
        txtFeedBack.setText(R.string.send_feedback);
        txtPrivacyPolicy.setText(R.string.privacy_policy);
        txtTermOfUse.setText(R.string.terms_of_use);
        txtSignOut.setText(R.string.sign_out);

    }
}

File: app/src/main/java/com/iceteaviet/fastfoodfinder/ui/settings/SettingActivity.kt
package com.iceteaviet.fastfoodfinder.ui.settings

import android.content.res.Configuration
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.widget.SwitchCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import dagger.hilt.android.AndroidEntryPoint
import com.iceteaviet.fastfoodfinder.R
import com.iceteaviet.fastfoodfinder.databinding.ActivitySettingBinding
import com.iceteaviet.fastfoodfinder.ui.base.BaseActivity
import com.iceteaviet.fastfoodfinder.ui.settings.discountnotify.DiscountNotifyDialog
import com.iceteaviet.fastfoodfinder.utils.openLoginActivity
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.util.Locale

@AndroidEntryPoint
class SettingActivity : BaseActivity() {

    private val viewModel: SettingViewModel by viewModels()

    private lateinit var binding: ActivitySettingBinding

    private lateinit var txtShareApp: TextView
    private lateinit var txtChangeMetric: TextView
    private lateinit var txtEditProfile: TextView
    private lateinit var txtChangePassword: TextView
    private lateinit var txtChangeEmail: TextView
    private lateinit var txtSetNotification: TextView
    private lateinit var txtSetEmailNotification: TextView
    private lateinit var layoutUpdateDb: LinearLayout
    private lateinit var progressBarUpdateDb: ProgressBar
    private lateinit var imageUpdateDb: ImageView
    private lateinit var txtAboutApp: TextView
    private lateinit var txtRateApp: TextView
    private lateinit var txtFeedBack: TextView
    private lateinit var txtPrivacyPolicy: TextView
    private lateinit var txtTermOfUse: TextView
    private lateinit var txtSignOut: TextView
    private lateinit var swChangeLanguage: SwitchCompat
    private lateinit var tvSettingLanguage: TextView

    override val layoutId: Int
        get() = R.layout.activity_setting

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivitySettingBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupUI()
        setupEventListeners()
        setupObservers()
    }

    override fun onResume() {
        super.onResume()
        viewModel.start()
    }

    private fun setupObservers() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collectLatest { state ->
                    swChangeLanguage.isChecked = !state.isVietnamese

                    if (state.showSignOutButton) {
                        txtSignOut.visibility = View.VISIBLE
                    } else {
                        txtSignOut.visibility = View.INVISIBLE
                    }

                    imageUpdateDb.visibility = if (state.showLoadingProgressIndicator) View.GONE else View.VISIBLE
                    progressBarUpdateDb.visibility = if (state.showLoadingProgressIndicator) View.VISIBLE else View.GONE

                    when (state.event) {
                        is SettingEvent.Idle -> {}
                        is SettingEvent.LoadLanguage -> {
                            loadLanguage(state.event.languageCode)
                            viewModel.markEventConsumed()
                        }
                        is SettingEvent.OpenLogin -> {
                            openLoginActivity(this@SettingActivity)
                            finish()
                            viewModel.markEventConsumed()
                        }
                        is SettingEvent.ShowSuccessLoadingToast -> {
                            Toast.makeText(this@SettingActivity, getString(R.string.update_database_successfull) + state.event.message, Toast.LENGTH_SHORT).show()
                            viewModel.markEventConsumed()
                        }
                        is SettingEvent.ShowFailedLoadingToast -> {
                            Toast.makeText(this@SettingActivity, getString(R.string.update_database_failed) + state.event.message, Toast.LENGTH_SHORT).show()
                            viewModel.markEventConsumed()
                        }
                    }
                }
            }
        }
    }

    private fun loadLanguage(languageToLoad: String) {
        val locale = Locale(languageToLoad)
        Locale.setDefault(locale)
        val configuration = Configuration()
        configuration.setLocale(locale)
        baseContext.resources.updateConfiguration(configuration, baseContext.resources.displayMetrics)
        refreshUI()
    }

    private fun setupUI() {
        findViews()
    }

    private fun findViews() {
        txtShareApp = binding.tvSettingShareApp
        txtChangeMetric = binding.tvSettingChangeMetric
        txtEditProfile = binding.tvSettingEditProfile
        txtChangePassword = binding.tvSettingChangePassword
        txtChangeEmail = binding.tvSettingChangeEmail
        txtSetNotification = binding.tvSettingNotification
        txtSetEmailNotification = binding.tvSettingNotificationEmail
        layoutUpdateDb = binding.llSettingUpdateDb
        progressBarUpdateDb = binding.progressBarUpdateDb
        imageUpdateDb = binding.ivUpdateDb
        txtAboutApp = binding.tvSettingAboutApp
        txtRateApp = binding.tvSettingRateApp
        txtFeedBack = binding.tvSettingFeedback
        txtPrivacyPolicy = binding.tvSettingPrivacyPolicy
        txtTermOfUse = binding.tvSettingTermOfUse
        txtSignOut = binding.tvSettingSignOut
        swChangeLanguage = binding.swLanguages
        tvSettingLanguage = binding.tvSettingEnglish
    }

    private fun refreshUI() {
        txtShareApp.setText(R.string.share_app_with_friends)
        txtChangeMetric.setText(R.string.use_metric_units)
        txtEditProfile.setText(R.string.edit_your_profile)
        txtChangePassword.setText(R.string.change_your_password)
        txtChangeEmail.setText(R.string.change_your_email)
        txtSetNotification.setText(R.string.set_notifications)
        tvSettingLanguage.setText(R.string.english)
        txtAboutApp.setText(R.string.about_fastfood_finder)
        txtRateApp.setText(R.string.rate_app)
        txtFeedBack.setText(R.string.send_feedback)
        txtPrivacyPolicy.setText(R.string.privacy_policy)
        txtTermOfUse.setText(R.string.terms_of_use)
        txtSignOut.setText(R.string.sign_out)
    }

    private fun setupEventListeners() {
        txtSignOut.setOnClickListener {
            viewModel.onSignOutClicked()
        }

        swChangeLanguage.setOnClickListener {
            viewModel.onLanguageChanged()
        }

        tvSettingLanguage.setOnClickListener {
            viewModel.onLanguageChanged()
        }

        txtSetNotification.setOnClickListener {
            val dlg = DiscountNotifyDialog.newInstance()
            dlg.show(supportFragmentManager, "dialog-filter")
        }

        layoutUpdateDb.setOnClickListener {
            viewModel.onLoadStoreFromServer()
        }
    }
}